src/utils/userStatus.ts

```diff
@@ -0,0 +1,66 @@
+/**
+ * SPDX-FileCopyrightText: 2020 Nextcloud GmbH and Nextcloud contributors
+ * SPDX-FileCopyrightText: 2023 Nextcloud GmbH and Nextcloud contributors
+ * SPDX-License-Identifier: AGPL-3.0-or-later
+ */
+import { t } from '@nextcloud/l10n'
+
+import type {
+	Conversation,
+	Participant,
+	ParticipantSearchResult,
+	ParticipantStatus,
+} from '../types'
+
+type AnyDataWithUserStatus = Conversation | Participant | ParticipantSearchResult
+
+/**
+ * Parse user status data from conversation, participant or search result to a single type object
+ * @param userData current user data
+ */
+function parseUserStatus(userData: AnyDataWithUserStatus): ParticipantStatus {
+	if (typeof userData.status === 'object') {
+		return userData.status as ParticipantStatus
+	} else {
+		return {
+			status: (userData as Participant).status,
+			message: (userData as Participant).statusMessage,
+			icon: (userData as Participant).statusIcon,
+			clearAt: (userData as Participant).statusClearAt,
+		}
+	}
+}
```

Comment:
Using so many type assertions makes TS not useful because it doesn't check anything anymore

Reply:
For example, it allows the function with `ParticipantSearchResult`.

ParticipantSearchResult has `status: ParticipantStatus | string`.

So it does to else and `status` can be `string`. Type assertion hides this case. Why? Is it an issue and a lost case, or there is something hidden that guarantees that `ParticipantSearchResult` has `status` only as `ParticipantStatus` but not string?

Reply:
We have this JS check in component for similar purposes:
```js
if (Object.prototype.hasOwnProperty.call(this.participant, 'statusMessage')) {
	// We preloaded the status when via participants API
	return {
		status: this.participant.status || null,
		message: this.participant.statusMessage || null,
		icon: this.participant.statusIcon || null,
	}
}
if (Object.prototype.hasOwnProperty.call(this.participant, 'status')) {
	// We preloaded the status when via search API
	return {
		status: this.participant.status.status || null,
		message: this.participant.status.message || null,
		icon: this.participant.status.icon || null,
	}
}
```
I wonder if it's better to prepare an object for TS util before calling it?

Reply:
The question is, is it actually guaranteed (by some logic) that here function is called only with `ParticipantSearchResult` with object status and never with the string one?

For example, if we know when `ParticipantSearchResult` has object status and when string.

If yes - we should adjust the type.

If no - we need to add an explicit check like `'statusMessage' in userData`.

Reply:
Yes, we have:
- status: string (offline, dnd, away) from partcipiants API / conversation API (statusMessage is present)
- status: object from search API (for users) (status.message is present)
- status string (empty '') from search API (e.g. for groups) (no other fields)

Reply:
> * status string (empty '') from search API (e.g. for groups) (no other fields)

But `ParticipantSearchResult` in new type says `status: ParticipantStatus | string,`

Reply:
then I'll change to ''